crates/ra_assists/src/handlers/create_function.rs

```diff
@@ -0,0 +1,313 @@
+use ra_syntax::{
+    ast::{self, AstNode},
+    SmolStr, SyntaxKind, SyntaxNode, TextUnit,
+};
+
+use crate::{Assist, AssistCtx, AssistId};
+use ast::{CallExpr, Expr};
+use ra_fmt::leading_indent;
+
+// Assist: create_function
+//
+// Creates a stub function with a signature matching the function under the cursor.
+//
+// ```
+// fn foo() {
+//      bar<|>("", baz());
+// }
+//
+// ```
+// ->
+// ```
+// fn foo() {
+//      bar("", baz());
+// }
+//
+// fn bar(arg_1: &str, baz: Baz) {
+//      todo!();
+// }
+//
+// ```
+pub(crate) fn create_function(ctx: AssistCtx) -> Option<Assist> {
+    let path: ast::Path = ctx.find_node_at_offset()?;
+
+    if ctx.sema.resolve_path(&path).is_some() {
+        // The function call already resolves, no need to create a function
+        return None;
+    }
+
+    if path.qualifier().is_some() {
+        return None;
+    }
+
+    let call: ast::CallExpr = ctx.find_node_at_offset()?;
+
+    let (generated_fn, cursor_pos, text_start) = generate_fn(&ctx, &call)?;
+
+    ctx.add_assist(AssistId("create_function"), "Create function", |edit| {
+        edit.target(call.syntax().text_range());
+
+        edit.set_cursor(cursor_pos);
+        edit.insert(text_start, generated_fn);
+    })
+}
+
+/// Generates a function definition that will allow `call` to compile.
+/// The function name must match.
+/// The arguments must have valid, nonconflicting names.
+/// The arguments' types must match those being passed to `call`.
+/// TODO generics?
+fn generate_fn(ctx: &AssistCtx, call: &CallExpr) -> Option<(String, TextUnit, TextUnit)> {
+    let (start, indent) = next_space_for_fn(&call)?;
+    let indent = if let Some(i) = &indent { i.as_str() } else { "" };
+    let mut fn_buf = String::with_capacity(128);
+
+    fn_buf.push_str("\n\n");
+    fn_buf.push_str(indent);
+    fn_buf.push_str("fn ");
+
+    let fn_name = fn_name(&call)?;
+    fn_buf.push_str(&fn_name);
+
+    let fn_generics = fn_generics(&call)?;
```

Comment:
Rather than calling `fn_generics` in the middle of generation, I'd rather use an explicitelly two-step API:

```
struct FunctionTemplate {
    ...
}

impl FunctionTemplate {
  fn from_call(call: &ast::CallExpr) -> Option<Self>;
  fn render(&self) -> (String, TextUnit, TextUnit);
}
```

Reply:
I like that approach, thanks for the suggestion!